pkg/ha/ha_service.go

```diff
@@ -0,0 +1,115 @@
+package ha
+
+import (
+	"context"
+	"fmt"
+	"golang.org/x/sync/semaphore"
+	"sync"
+	"time"
+
+	"github.com/timescale/promscale/pkg/log"
+	"github.com/timescale/promscale/pkg/pgxconn"
+)
+
+const backOffDurationOnLeaderChange = 10 * time.Second
+
+type Service struct {
+	state             *State
+	lockClient        haLockClient
+	leaseTimeout      time.Duration
+	leaseRefresh      time.Duration
+	_leaderChangeLock *semaphore.Weighted
+}
+
+func NewHAService(dbClient pgxconn.PgxConn) (*Service, error) {
+	lockClient := newHaLockClient(dbClient)
+	timeout, refresh, err := lockClient.readLeaseSettings(context.Background())
+	if err != nil {
+		return nil, fmt.Errorf("could not create HA state: %#v", err)
+	}
+	return &Service{
+		state: &State{
+			_mu: sync.RWMutex{},
+		},
+		lockClient:   lockClient,
+		leaseTimeout: timeout,
+		leaseRefresh: refresh,
+	}, nil
+}
+
+// checkInsert verifies the samples are from prom leader & in expected time range.
+// 	returns true if the insert is allowed
+func (h *Service) checkInsert(minT, maxT time.Time, clusterName, replicaName string) (bool, error) {
+	// TODO use proper context
+	lockState, err := h.lockClient.checkInsert(context.Background(), clusterName, replicaName, minT, maxT)
+
+	// unknown error while checking lock state
+	if err != nil && err != leaderHasChanged {
+		return false, fmt.Errorf("could not check ha lock state: %#v", err)
+	}
+
+	// leader changed
+	if err == leaderHasChanged {
+		// read latest lock state
+		lockState, err = h.lockClient.readLockState(context.Background(), clusterName)
+		// couldn't get latest lock state
+		if err != nil {
+			return false, fmt.Errorf("could not check ha lock state: %#v", err)
+		}
+		// update state asynchronously
+		go h.state.update(lockState, replicaName, maxT)
```

Comment:
We can definitely change this to be synchronous. But in the current flow, the conditions we are evaluating aren't dependent on the updation because we are explicitly using `lockState`, `replicaName`, & `maxT` in the next checks.